xmake/rules/swift/xmake.lua

```diff
@@ -18,15 +18,182 @@
 -- @file        xmake.lua
 --
 
+rule("swift.interop", function()
+    set_sourcekinds("sc")
+    add_orders("swift.interop", "swift.build")
+    add_orders("swift.interop", "c++.build")
+    on_config(function(target)
+
+        import("core.base.json")
+
+        if not target:values("swift.interop") then
+            return
+        end
+
+        local sourcebatches = target:sourcebatches()
+        local sourcebatch = sourcebatches and sourcebatches["swift.interop"]
+        local sourcefiles = sourcebatch and sourcebatch.sourcefiles
+        sourcefiles = sourcefiles or {}
+        local enabled = false
+        for _, sourcefile in ipairs(sourcefiles) do
+            local fileconfig = target:fileconfig(sourcefile)
+            if fileconfig and fileconfig.public then
+                enabled = true
+                break
+            end
+        end
+        if not enabled then
+            return
+        end
+
+        local sc = target:tool("sc")
+        assert(sc, "No swift compiler found!")
+        local outdata, errdata = os.iorunv(sc, {"-print-target-info"})
+        assert(outdata, errdata)
+
+        local target_info = json.decode(outdata)
+        assert(target_info and target_info.paths and target_info.paths.runtimeResourcePath, "Failed to get swift resource path")
+        target:add("includedirs", target_info.paths.runtimeResourcePath, {public = true})
+
+        local outdir = target:autogendir("swift.interop")
+        target:add("includedirs", outdir, {public = true})
+
+        local mode = (type(target:values("swift.interop")) == "string") and target:values("swift.interop") or "objc"
+
+        local headername = (target:values("swift.interop.headername") or (target:name() .. "-Swift.h"))
+        local header = path.join(outdir, headername)
+        target:add("headerfiles", header)
+
+        target:data_set("swift.interop", mode)
+        target:data_set("swift.interop.outdir", outdir)
+    end)
+
+    on_prepare_files(function(target, jobgraph, sourcebatch, opt)
+
+        if not target:data("swift.interop") then
```

Comment:
if we use target:rule_enable to disable rule, on_prepare_files will not be called.